### PR TITLE
SV-330 Filter Implementation & SV-278 Lucene Fuzzy Match Updates

### PR DESCRIPTION
**Pull request by CodeWolf86**

### CODE REVIEW

File: src/SFA.DAS.Courses.Data/Configuration/Standard.cs [hunk 1]
```diff
@@ -41,6 +41,7 @@ public void Configure(EntityTypeBuilder<Domain.Entities.Standard> builder)
                 .WithMany(c => c.Standards)
                 .HasForeignKey(s => s.LarsCode)
                 .HasPrincipalKey(l => l.LarsCode)
+                .IsRequired(false)
```
Comment (CodeWolf86) [inline src/SFA.DAS.Courses.Data/Configuration/Standard.cs:44]: Is Required False changes how the join is added when performing an Include. Ie if a Lars Record doesn't exist ( as they don't with in development standards ) then they still get returned. If this line is omitted, those records aren't returned when no filter is set.

**Review by DevenKShah — APPROVED** [review]
This looks good to me.

File: src/SFA.DAS.Courses.Data/Extensions/StandardFilterExtension.cs [hunk 2]
```diff
@@ -9,6 +12,8 @@ public static class StandardFilterExtension
     {
         public static IQueryable<Standard> FilterStandards(this IQueryable<Standard> standards, StandardFilter filter)
         {
+            // Active and ActiveAvailable Filters require IsLatestVersion filter, but the query is to complex for EF
```
[the remark this answers is not in the document]
Comment (CodeWolf86) [reply src/SFA.DAS.Courses.Data/Extensions/StandardFilterExtension.cs:15]: Sorted 👍

File: src/SFA.DAS.Courses.Api.UnitTests/Controllers/Standards/WhenCallingGetStandardsList.cs [hunk 3]
```diff
@@ -38,7 +38,7 @@ public class WhenCallingGetStandardsList
                         query.RouteIds.Equals(routeIds) &&
                         query.Levels.Equals(levels) &&
                         query.OrderBy.Equals(orderBy) &&
-                        query.Filter.Equals(StandardFilter.ActiveAvailable)), 
+                        query.Filter.Equals(StandardFilter.None)),
```
[the remark this answers is not in the document]
Comment (CodeWolf86) [reply src/SFA.DAS.Courses.Api.UnitTests/Controllers/Standards/WhenCallingGetStandardsList.cs:41]: Have removed the comment on changed to specify the filter to none and pass into the mock

File: src/SFA.DAS.Courses.Data/Extensions/StandardFilterExtension.cs [hunk 4]
```diff
@@ -17,28 +22,74 @@ public static IQueryable<Standard> FilterStandards(this IQueryable<Standard> sta
...
+                    standards = standards.FilterNotYetApproved();
+                    break;
                 case StandardFilter.None:
                     break;
             }
 
             return standards;
         }
 
+        public static IEnumerable<Standard> InMemoryFilterIsLatestVersion(this IEnumerable<Standard> standards, StandardFilter filter)
```
[the remark this answers is not in the document]
Comment (CodeWolf86) [reply src/SFA.DAS.Courses.Data/Extensions/StandardFilterExtension.cs:35]: Have talked to @DevenKShah  about options here. Going to merge as is to unblock other tickets and raise a ticket to potentially rework this area See ticket https://skillsfundingagency.atlassian.net/browse/SV-401

File: src/SFA.DAS.Courses.Data/Repository/StandardRepository.cs [hunk 5]
```diff
@@ -79,10 +79,13 @@ public async Task<IEnumerable<Standard>> GetFilteredStandards(IList<Guid> routeI
...
-                .ThenInclude(c=>c.SectorSubjectArea);
-                
+                .ThenInclude(c => c.SectorSubjectArea);
 
-            return await standards.ToListAsync();
+            var standardResults = await standards.ToListAsync();
+
+            // Secondary filter performed in memory due to limitations in 
+            // EF Core query translation on Group By selecting top row of each.
+            return standardResults.InMemoryFilterIsLatestVersion(filter);
```
[the remark this answers is not in the document]
Comment (CodeWolf86) [reply src/SFA.DAS.Courses.Data/Repository/StandardRepository.cs:88]: Have talked to @DevenKShah  about options here. Going to merge as is to unblock other tickets and raise a ticket to potentially rework this area- See ticket SV-401 https://skillsfundingagency.atlassian.net/browse/SV-401

**Pull request merged**